File: 17 robots - multileg strangle hedge/cpp/strangle_hedge.cpp
#include "tse_helpers.hpp"

#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <fstream>
#include <string>
#include <vector>

namespace {

	std::int64_t constexpr enterCheckPoint {1000000000LL};
	std::int64_t constexpr exitCheckPoint {2000000000LL};
	std::int64_t constexpr hugeCoolDown {1000000000000000LL};

	struct LegSpec final {
		std::string symbol;
		tse::Instrument instrument;
		double quantity;
		tse::Side entrySide;
		double entryPrice;
		double exitPrice;
	};

	tse::Side flipSide(tse::Side const side)
	{
		if (side == tse::Side::long_) {
			return tse::Side::short_;
		}
		if (side == tse::Side::short_) {
			return tse::Side::long_;
		}
		return tse::Side::neutral;
	}

	std::vector<std::vector<double>> loadPriceRows(std::string const& name)
	{
		std::ifstream file {helpers::dataPath(name)};
		std::vector<std::vector<double>> rows;
		std::string line;
		std::getline(file, line);
		while (std::getline(file, line)) {
			if (line.empty()) {
				continue;
			}
			std::vector<double> row;
			std::size_t
				start {0},
				column {0};
			while (start <= line.size()) {
				std::size_t const sep {line.find(',', start)};
				std::string const cell {line.substr(start, sep == std::string::npos ? std::string::npos : sep - start)};
				if (column > 0) {
					row.push_back(std::stod(cell));
				}
				++column;
				if (sep == std::string::npos) {
					break;
				}
				start = sep + 1;
			}
			rows.push_back(row);
		}
		return rows;
	}

	void loadLegPrices
	(
		std::string const& name,
		std::vector<LegSpec>& legs
	)
	{
		std::vector<std::vector<double>> const rows {loadPriceRows(name)};
		for (std::size_t i {0}; i < legs.size(); ++i) {
			legs[i].entryPrice = rows[0][i];
			legs[i].exitPrice = rows[1][i];
		}
	}

	tse::TradeInputProcessor priceProcessor()
	{
		return []
		(
			tse::Storage const& storage,
			std::string const&,
			tse::TradeTick const& tick
		) -> bool
		{
			storage.push(tick.tsNanoseconds, tick.price);
			return true;
		};
	}

	tse::LegDescriptor makeLeg
	(
		std::string symbol,
		double const quantity,
		double const limitPrice,
		tse::TxnType const txnType,
		tse::Side const txnSide,
		tse::Side const posSide
	)
	{
		return tse::LegDescriptor
		{
			std::move(symbol),
			tse::QuantityMode::fixed,
			quantity,
			tse::PriceType::limit,
			limitPrice,
			0.0,
			0.0,
			txnType,
			txnSide,
			posSide,
			tse::Tif::day,
			tse::Priority::replaceable
		};
	}

	std::vector<tse::LegDescriptor> entryLegs(std::vector<LegSpec> const& legs)
	{
		std::vector<tse::LegDescriptor> out;
		out.reserve(legs.size());
		for (LegSpec const& leg : legs) {
			out.push_back(makeLeg(leg.symbol, leg.quantity, leg.entryPrice, tse::TxnType::enter, leg.entrySide, tse::Side::neutral));
		}
		return out;
	}

	std::vector<tse::LegDescriptor> exitLegs(std::vector<LegSpec> const& legs)
	{
		std::vector<tse::LegDescriptor> out;
		out.reserve(legs.size());
		for (LegSpec const& leg : legs) {
			out.push_back(makeLeg(leg.symbol, leg.quantity, leg.exitPrice, tse::TxnType::exit, flipSide(leg.entrySide), leg.entrySide));
		}
		return out;
	}

	void pushPrice
	(
		tse::Market const& market,
		std::string const& symbol,
		std::int64_t const tsNanoseconds,
		double const price
	)
	{
		market.pushTrade(symbol, tse::TradeTick {tsNanoseconds, price, 1.0, tse::Side::trade});
	}

}

int main()
{
	tse::setLogLevel(tse::LogLevel::none);

	// The future carries the directional view, the two puts underneath insure it against a fall
	// and the far call buys back the upside: four legs, one hedged position.
	std::vector<LegSpec> legs
	{
		LegSpec {"CLF26", tse::Instrument::future, 1.0, tse::Side::long_, 0.0, 0.0},
		LegSpec {"LO F26 P00045000", tse::Instrument::option, 1.0, tse::Side::long_, 0.0, 0.0},
		LegSpec {"LO F26 P00044000", tse::Instrument::option, 1.0, tse::Side::long_, 0.0, 0.0},
		LegSpec {"LO F26 C00060000", tse::Instrument::option, 1.0, tse::Side::long_, 0.0, 0.0}
	};
	loadLegPrices("multileg_wti_strangle.csv", legs);

	tse::Account account {"WtiStrangle", tse::StorageRegime::mem};
	tse::Market const market {account.createMarket("price", tse::MdType::trade)};
	account.createSimulator("sim", helpers::simulatorConfig(), 64, -1);
	// A WTI contract is a thousand barrels, so a dollar on the future is a thousand dollars of P&L.
	for (LegSpec const& leg : legs) {
		account.addContract(leg.symbol, 1000, leg.instrument, tse::Underlying::commodity, tse::Venue::CME, 100000);
	}
	account.addInputTrade("Px", 2, tse::Duration::nanoseconds, priceProcessor(), market, {legs[0].symbol});
	account.addPatternTimestamp("EnterAt", tse::Duration::nanoseconds, {"Px"}, enterCheckPoint, hugeCoolDown);
	account.addPatternTimestamp("ExitAt", tse::Duration::nanoseconds, {"Px"}, exitCheckPoint, hugeCoolDown);
	// A multileg transaction is all or nothing: the four legs fill at one instant against
	// a single net price, or none of them fills, so the hedge can never go on half-built.
	account.addRuleMultileg("Strangle enter", entryLegs(legs), "EnterAt");
	account.addRuleMultileg("Strangle exit", exitLegs(legs), "ExitAt");
	account.addRobot("WTI_STRANGLE", {"Strangle enter", "Strangle exit"});
	for (LegSpec const& leg : legs) {
		account.portfolioSubscribe(market, leg.symbol);
	}
	account.start("WTI_STRANGLE");

	pushPrice(market, legs[0].symbol, enterCheckPoint, legs[0].entryPrice);
	pushPrice(market, legs[0].symbol, 1000000110LL, legs[0].entryPrice);
	pushPrice(market, legs[1].symbol, 1000000120LL, legs[1].entryPrice);
	pushPrice(market, legs[2].symbol, 1000000130LL, legs[2].entryPrice);
	pushPrice(market, legs[3].symbol, 1000000140LL, legs[3].entryPrice);

	pushPrice(market, legs[0].symbol, exitCheckPoint, legs[0].exitPrice);
	pushPrice(market, legs[0].symbol, 2000000200LL, legs[0].exitPrice);
	pushPrice(market, legs[1].symbol, 2000000200LL, legs[1].exitPrice);
	pushPrice(market, legs[2].symbol, 2000000200LL, legs[2].exitPrice);
	pushPrice(market, legs[3].symbol, 2000000200LL, legs[3].exitPrice);

	tse::Summary const summary {account.getSummary()};

	std::printf("strangle_hedge: netProfit=%.2f\n", summary.totalNetProfit);
	return 0;
}

File: 17 robots - multileg strangle hedge/python/strangle_hedge.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

tse = H.tse

ENTER_CHECKPOINT = 1000000000
EXIT_CHECKPOINT = 2000000000
HUGE_COOLDOWN = 1000000000000000


def flip_side(side):
    if side == tse.Side.Long:
        return tse.Side.Short
    if side == tse.Side.Short:
        return tse.Side.Long
    return tse.Side.Neutral


def load_price_rows(name):
    rows = []
    with open(H.data_path(name)) as handle:
        handle.readline()
        for line in handle:
            line = line.strip()
            if not line:
                continue
            cells = line.split(",")
            rows.append([float(cell) for cell in cells[1:]])
    return rows


def load_leg_prices(name, legs):
    rows = load_price_rows(name)
    for i, leg in enumerate(legs):
        leg["entry_price"] = rows[0][i]
        leg["exit_price"] = rows[1][i]


def price_processor(storage, contract_id, tick):
    storage.push(tick.tsNanoseconds, tick.price)
    return True


def make_leg(symbol, quantity, limit_price, txn_type, txn_side, pos_side):
    return tse.make_leg_descriptor(
        symbol,
        tse.Quantity.Fixed,
        quantity,
        tse.Price.Limit,
        limit_price,
        0.0,
        0.0,
        txn_type,
        txn_side,
        pos_side,
        tse.Tif.Day,
        tse.Priority.Replaceable,
    )


def entry_legs(legs):
    return [
        make_leg(leg["symbol"], leg["quantity"], leg["entry_price"], tse.TxnType.Enter, leg["entry_side"], tse.Side.Neutral)
        for leg in legs
    ]


def exit_legs(legs):
    return [
        make_leg(leg["symbol"], leg["quantity"], leg["exit_price"], tse.TxnType.Exit, flip_side(leg["entry_side"]), leg["entry_side"])
        for leg in legs
    ]


def push_price(market, symbol, ts_nanoseconds, price):
    market.push_trade_by_name(symbol, tse.TseTickTrade(ts_nanoseconds, price, 1.0, int(tse.Side.Trade)))


def main():
    # The future carries the directional view, the two puts underneath insure it against a fall
    # and the far call buys back the upside: four legs, one hedged position.
    legs = [
        {"symbol": "CLF26", "instrument": tse.Instrument.Future, "quantity": 1.0, "entry_side": tse.Side.Long, "entry_price": 0.0, "exit_price": 0.0},
        {"symbol": "LO F26 P00045000", "instrument": tse.Instrument.Option, "quantity": 1.0, "entry_side": tse.Side.Long, "entry_price": 0.0, "exit_price": 0.0},
        {"symbol": "LO F26 P00044000", "instrument": tse.Instrument.Option, "quantity": 1.0, "entry_side": tse.Side.Long, "entry_price": 0.0, "exit_price": 0.0},
        {"symbol": "LO F26 C00060000", "instrument": tse.Instrument.Option, "quantity": 1.0, "entry_side": tse.Side.Long, "entry_price": 0.0, "exit_price": 0.0},
    ]
    load_leg_prices("multileg_wti_strangle.csv", legs)

    account = H.account("WtiStrangle", tse.StorageRegime.Mem, tse.LogLevel.Off)
    market = account.create_market("price", tse.MdType.Trade)
    account.create_simulator("sim", H.simulator_options(), 64, -1)
    # A WTI contract is a thousand barrels, so a dollar on the future is a thousand dollars of P&L.
    for leg in legs:
        account.add_contract(leg["symbol"], 1000, leg["instrument"], tse.Underlying.Commodity, tse.Venue.CME, 100000)
    account.add_input_trade("Px", 2, tse.Duration.Nanoseconds, price_processor, market, [legs[0]["symbol"]])
    account.add_pattern_timestamp("EnterAt", tse.Duration.Nanoseconds, ["Px"], ENTER_CHECKPOINT, HUGE_COOLDOWN)
    account.add_pattern_timestamp("ExitAt", tse.Duration.Nanoseconds, ["Px"], EXIT_CHECKPOINT, HUGE_COOLDOWN)
    # A multileg transaction is all or nothing: the four legs fill at one instant against
    # a single net price, or none of them fills, so the hedge can never go on half-built.
    account.add_rule_multileg("Strangle enter", entry_legs(legs), "EnterAt")
    account.add_rule_multileg("Strangle exit", exit_legs(legs), "ExitAt")
    account.add_robot("WTI_STRANGLE", ["Strangle enter", "Strangle exit"])
    for leg in legs:
        account.portfolio_subscribe(market, leg["symbol"])
    account.start("WTI_STRANGLE")

    push_price(market, legs[0]["symbol"], ENTER_CHECKPOINT, legs[0]["entry_price"])
    push_price(market, legs[0]["symbol"], 1000000110, legs[0]["entry_price"])
    push_price(market, legs[1]["symbol"], 1000000120, legs[1]["entry_price"])
    push_price(market, legs[2]["symbol"], 1000000130, legs[2]["entry_price"])
    push_price(market, legs[3]["symbol"], 1000000140, legs[3]["entry_price"])

    push_price(market, legs[0]["symbol"], EXIT_CHECKPOINT, legs[0]["exit_price"])
    push_price(market, legs[0]["symbol"], 2000000200, legs[0]["exit_price"])
    push_price(market, legs[1]["symbol"], 2000000200, legs[1]["exit_price"])
    push_price(market, legs[2]["symbol"], 2000000200, legs[2]["exit_price"])
    push_price(market, legs[3]["symbol"], 2000000200, legs[3]["exit_price"])

    summary = account.get_summary()
    account.close()

    print("strangle_hedge: netProfit={:.2f}".format(summary.totalNetProfit))
    return 0


if __name__ == "__main__":
    sys.exit(main())
